kep/61-real-load-aware-scheduling/README.md

```diff
@@ -0,0 +1,572 @@
+
+<img src="images/design.png" alt="design" width="720" height="628"/>
+
+
+
+### Metrics Provider
+
+This service provides metrics backed by a time-series database — for example, Prometheus, InfluxDB, [Kubernetes Metrics Server](https://github.com/kubernetes-sigs/metrics-server) etc.
+
+
+### Load Watcher/Load Analyser
+
+The load watcher and analyzer both run in a single process. The watcher is responsible for retrieving cluster-wide resource usage metrics like CPU, memory, network, and IO stats over windows of a specified duration from metrics providers above. It stores these in its local cache and persists several aggregations in the host local DB for fault tolerance. The analyzer is responsible for the detection of bad metrics and any remediation. Bad metrics could be due to missing metrics or metrics with considerable errors, making them anomalies. It can also be extended in the future to use ML models for analyzing metrics.
+
+Load watcher would cache metrics in the last 15-minute, 10-minute, and 5-minute windows, which can be queried via REST API exposed. A generic interface will be provided that can support fetching metrics from any metrics provider. Example is provided in Appendix.
+
+
+### DB
+
+This is a localhost database stored as a file. This DB aims to populate the load watcher's cache quickly if it is lost due to a crash and perform scheduling decisions without any impact on latency.
+The file will be stored in host file system, so it will be persisted across pod failures on the same node. For HA setups, each load watcher process will maintain its corresponding file locally.
+
+
+### Scheduler Plugins
+
+This uses the scheduler framework of K8s to incorporate our customized real load aware scheduler plugins without modifying the core scheduler code. The plugins we proposed mainly include the following two.
+
+- BestFitBinPack Plugin: It is best fit variant of bin pack algorithm that scores nodes by their actual resource utilization in a way that all utilized nodes have around x% of utilization.
+- Safe Balancing Plugin: It is a node sorting plugin that sorts nodes base on both the mean and the standard deviation of node resource utilization. It aims to balance not only the average load but also the risk caused by load variations.
+
+
+### BestFitBinPack Plugin
+
+
+#### Score plugin
+
+This plugin would extend the Score extension point. K8s scheduler framework calls the Score function for each node separately when scheduling a pod.
+
+Following is the algorithm:
+
+**Algorithm**
+
+1. Get the utilization of the current node to be scored. Call it A.
+2. Calculate the current pod's total CPU requests and overhead. Call it B.
+3. Calculate the expected utilization if the pod is scheduled under this node by adding i.e. U = A + B.
+4. If U &lt;= X%, return U+X as the score
```

Comment:
say X = 70 and U = 60, then limits don't work with these formulas.

And again, I don't think I heard back about the discontinuous function. Did you get to try a continuous function?
It would be interesting to see which formula gives you better utilization over a period of time.

By continuous I mean something like: 50 to 100 score from 0 to 50% utilization. Then 100 to 0 score from 50 to 100% utilization

Reply:
Yes you are right, X=50% is a special use case per our requirements, since I recently changed the algorithm to generalize I missed this part. I will add the correct generalized equations. Good catch.

The discontinuity here can be thought of as a penalty for nodes going above target utilisation as we want to be aggressive in keeping the node under target when possible, and then move from best fit to least fit variant of bin packing if no nodes are available with < target utilisation. A break in continuity is needed to incorporate penalty in scoring function. The continuous function you mentioned is closer to the target, which we want to avoid. Based on this reasoning we didn't try a continuous function, and this function works good per experiments and simulations.

Reply:
I understand you want to penalize the node that has utilization larger than X%, but @alculquicondor 's point is that a node with utilization of 51% should be penalized that much, comparing to a node with 49% utilization - the score drops from nearly 100 to 50.

So, does it make sense to update the formula on the range of [50%, 100%] to be `2•(100-U)` when X% is 50%? And a general formula is like: `100•(100-U) / (100-X)`.

Reply:
"utilization of 51% should be penalized that much, comparing to a node with 49% utilization"
Better way to put this is we want to "favor" nodes <= 50% utilization and "penalise" nodes with > 50% utilisation.

With 2*(100-U) the problem is that we will score node with utilization 40 and 65 to be same i.e. 90. This would go against the constraint we have as we have equal chance of selecting underutilised and overutilised nodes. Hope this example makes it clear for the need of a break.

Reply:
> With 2*(100-U) the problem is that we will score node with utilization 40 and 65 to be same i.e. 90.

How about using a logarithmic formula to ensure node with utilization 65% gets scored less than 40%?

Reply:
Hold on... Isn't 40% gets a Score of 90, and 65% gets a Score of 70?

Reply:
If the issue is about continuity, we _could_ explore a continuous non linear function for utilisation > 50%, and analyse how it compares to a break. We can leave it for a future research project.

Reply:
Yup sorry not 65 -> 55*

Reply:
> We can leave it for a future research project or enhancement.

Fair, it's a score function after all.

Reply:
I think It is too hard to have a perfect score function that works for everyone. We may have 2 separate score functions if we need but lets keep the version for now.